### DISCUSSION TITLE
Where is daedalOS still in the "Uncanny Valley of User Interface"

### DISCUSSION THREAD

**Discussion by DustinBrett** [DustinBrett/daedalOS#216] [Ideas]
In what ways does this app still feel not quite right?

https://blog.codinghorror.com/avoiding-the-uncanny-valley-of-user-interface/

**Comment by Joshfindit** [comment]
If you don’t mind me clarifying:

Are you asking about opportunities for it to feel more like a web app? Or are you wanting it to feel like a traditional windowed desktop and asking for ways that it feels “too much” like a web app?

**Comment by DustinBrett** [comment]
Thanks for asking @Joshfindit! I desire for it to feel more like a desktop environment. Ideally if someone came up to daedalOS already in full screen, they would hopefully have a hard time knowing they are on a web page. And I am interested in the point in which they realize they are not in Windows 10, so that I can adjust to make the illusion last longer.

**Comment by Joshfindit** [comment]
Excellent. I do have some thoughts in that direction:

- The animated background
  - I'm sad to say this, but the animated "Vanta Waves" background is just too smooth and the design is web-like
  - Windows has never had good support for animated backgrounds. While there are 3rd party tools that provide a better experience, I have yet to see one that's as cleanly implemented as [daedalOS](https://github.com/DustinBrett/daedalOS)'
  - I wouldn't set it as default, but standard "Windows blue" could be a good addition to the background choices
- Systray
  - Windows *always* has stuff in the systray
  - You might get away with just adding the "show hidden icons" button, but that just means it would be obvious when you click it and it doesn't show icons
- Clock
  - Universally, clicking the clock brings up the calendar. For this to pass as Windows I don't think anything further needs to be done besides highlighting whichever date is clicked
- Terminal
  - This is silly, but while people don't usually notice the difference between `/` and `\`, I think they would notice that the `C:` is missing
- The async loading of the icons
  - I was kind of surprised that this jumped out so much to me since it happens so quickly, but it does and there's probably a simple fix: hide them until they are all loaded, then start on previews. I think Windows does this, but it's been a minute since I've noticed the icons loading on a machine.
  - Note: I mean the icons themselves, and not the previews. async loading of previews looks just like it does on Windows. Especially when opening a folder of images
- One minor one: Windows caches the icon previews so navigating to a folder the second time feels instant. While some webapps do the same, it's much more common for a web app to force a full refresh

**Comment by Joshfindit** [comment]
The async loading of the icons.. It's hard to catch it since it moves so fast, but here are some additional thoughts:

- It might be that Windows loads a default icon first before loading the preview
- It might be that [daedalOS](https://github.com/DustinBrett/daedalOS) feeds the icons in to the folder window 'out of order'. The Windows Explorer process is very clearly single-threaded and so the icons always load in sequence

**Comment by DustinBrett** [comment]
Thanks for this feedback @Joshfindit. Sorry for the delay in responding.

- Animated Wallpaper: Thanks for considering the integration I did smooth & cleanly implemented. For this I do agree it's not the standard Windows look but I decided to go animated as I've always though they were cool and I took my inspiration from [Wallpaper Engine](https://store.steampowered.com/app/431960/Wallpaper_Engine/) & [Lively](https://rocksdanister.github.io/lively/) which both use basically the same thing as Vanta Waves in their main demos.

- System Tray: I absolutely plan to add this option and I've been waiting for a use-case to make the thing. I was thinking I could have an option in Webamp to allow minimizing to sys tray, which I used to do back in the olden days. That being said, I run 3 monitors and I hide the systray icons personally and on the 2 side monitors, it doesn't even show up at all, which is how I've emulated it on daedalOS.

- Clock: That's a good idea. I had planned eventually to make a calendar app but I hadn't thought to add the clock one. I feel like maybe early Windows 10 days there actually wasn't a calendar anymore and they took it away, but maybe I just forgot that it's always been there. I will make a note to add this.

- Terminal: Indeed not using C:\ is a bit different here, but I guess it's a decision I decided to make at a technical level to slightly have inspiration from nix and also to better match url paths. I don't think I will change this as not using C:\ was a choice I made early on.

- Icon Loading: Could you clarify further the exact visual issue, because I have spent some time on this icon loading and it MAY work slightly different on slower machines vs fast ones, also possibly the desktop may act different because I've done some things to allow the desktop to be special to load faster on initial page loads. I do in some sense have it load a default icon and then load the cached one or make a dynamic one, but it might happen too quick for the default one to load because even that takes time.

**Reply by Joshfindit** [reply]
> took my inspiration from [Wallpaper Engine](https://store.steampowered.com/app/431960/Wallpaper_Engine/) & [Lively](https://rocksdanister.github.io/lively/) which both use basically the same thing as Vanta Waves in their main demos.

I installed Lively and put the Waves wallpaper on. You're right: yours is very close. I don't know if my Windows is just set wrong (classic Windows problem, and kind of what I'm saying is on the 'realistic' side of the uncanny valley), but the taskbar does not show the background behind it, and certainly not with the glass-like blur you have in daedalOS.

**Reply by DustinBrett** [reply]
Thanks for checking! I never even did that myself tbh, just took inspiration visually and in code. I'm glad to hear mine looked good in comparison. I agree the wallpaper is probably going to make them question it's Windows 10, but I guess the main part of this uncanny valley post was to see if anyone noticed places where I tried to copy Windows 10 and was wrong or not quite right by a large enough margin for it to be jarring.

**Comment by Joshfindit** [comment]
> Icon Loading: Could you clarify further the exact visual issue ... but it might happen too quick for the default one to load because even that takes time.

This one is tricky. If I was closer to this project I'd set up a test environment and record at 120fps, but I *think* I can explain it this way:

In Windows explorer, the thumbnails/icons load in two stages:

1. The filenames and initial icons are loaded.
    This places everything on the screen and since the icons have a specific grid size there is no shift in the layout during this sequence or the next. This process is sequential and blocking.
2. The thumbnails are generated / loaded
    This one is a bit more loosy-goosy in recent versions of Windows due to `thumbs.db` and lazy-loading the images as they scroll in to view, but it's still mostly top-left to bottom-right.

I think you could do a couple things to make this look almost exactly the same as Windows:

1. Clear the window when navigating between folders
2. Have the process that builds the file listing pre-queue the standard icons and only start populating the window when the icons are ready. *Way back in (Win95? 3.1?) there was even a default icon for applications and the custom icons 'popped in' afterwards.*
3. Populate the filenames and default icons sequentially. No async
4. Once 3 is done, do a second pass and replace the default icons with thumbnails

It will likely make it slightly slower, but IMO it will mean that the file explorer jumps right over the uncanny valley

**Reply by DustinBrett** [reply]
Thanks for this detailed feedback, sorry for my slow reply. I have been looking into this but still working on getting ideal video/recreations of what exactly does happen in those milliseconds as File Explorer loads.

In general you are right that it's not exact and also it's slower than I'd like in some ways. So I will be working on file entry / icon loading going forward and will take these notes into consideration.
